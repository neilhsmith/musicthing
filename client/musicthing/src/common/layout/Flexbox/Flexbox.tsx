import { ReactNode, createElement } from "react";
import styled from "styled-components";

import { SpacingProps, generateSpacingCss } from "utils/spacing";
import { SizingProps, generateSizingCss } from "utils/sizing";

interface FlexboxProps extends SpacingProps, SizingProps {
  alignContent?:
    | "flex-start"
    | "flex-end"
    | "center"
    | "space-between"
    | "space-around"
    | "stretch";
  alignItems?: "flex-start" | "flex-end" | "center" | "baseline" | "stretch";
  alignSelf?: "flex-start" | "flex-end" | "center" | "baseline" | "stretch";
  children: ReactNode;
  display?: "flex" | "inline-flex";
  element?:
    | "article"
    | "aside"
    | "div"
    | "figure"
    | "footer"
    | "header"
    | "main"
    | "nav"
    | "section";
  flex?: number | string;
  flexBasis?: number | string;
  flexDirection?: "column-reverse" | "column" | "row-reverse" | "row";
  flexFlow?: string; // todo: limit this to actual flex-flow values
  flexGrow?: number | string;
  flexShrink?: number | string;
  flexWrap?: "nowrap" | "wrap-reverse" | "wrap";
  justifyContent?:
    | "center"
    | "flex-end"
    | "flex-start"
    | "space-around"
    | "space-between";
}

const Flexbox = styled(
  ({
    alignContent,
    alignItems,
    alignSelf,
    children,
    display,
    element,
    flex,
    flexBasis,
    flexDirection,
    flexFlow,
    flexGrow,
    flexShrink,
    flexWrap,
    justifyContent,
    margin,
    marginTop,
    marginRight,
    marginBottom,
    marginLeft,
    padding,
    paddingTop,
    paddingRight,
    paddingBottom,
    paddingLeft,
    height,
    maxHeight,
    minHeight,
    width,
    maxWidth,
    minWidth,
    ...otherProps
  }: FlexboxProps) => createElement(element || "div", otherProps, children)
)`
  ${(props) => propIfExists(props.alignContent, "align-content")}
  ${(props) => propIfExists(props.alignItems, "align-items")}
  ${(props) => propIfExists(props.alignSelf, "align-self")}
  ${(props) => propIfExists(props.display, "display") ?? "display: flex;"}
  ${(props) => propIfTruthyOrZero(props.flex, "flex")}
  ${(props) => propIfTruthyOrZero(props.flexBasis, "flex-basis")}
  ${(props) => propIfExists(props.flexDirection, "flex-direction")}
  ${(props) => propIfExists(props.flexFlow, "flex-flow")}
  ${(props) => propIfTruthyOrZero(props.flexGrow, "flex-grow")}
  ${(props) => propIfTruthyOrZero(props.flexShrink, "flex-shrink")}
  ${(props) => propIfExists(props.flexWrap, "flex-wrap")}
  ${(props) => propIfExists(props.justifyContent, "justify-content")}
  ${(props) => generateSpacingCss(props)}
  ${(props) => generateSizingCss(props)}
`;

export default Flexbox;

function propIfExists(prop: string | number | undefined, cssClass: string) {
  return prop ? `${cssClass}: ${prop};` : null;
}

function propIfTruthyOrZero(
  prop: number | string | undefined,
  cssClass: string
) {
  if (prop || prop === 0) {
    return propIfExists(prop, cssClass);
  }
  return null;
}
