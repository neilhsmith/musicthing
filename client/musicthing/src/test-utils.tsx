import React, { ReactElement } from "react";
import { render, screen } from "@testing-library/react";
import { ThemeProvider } from "styled-components";

import { dark } from "styles/themes";

const AllProviders = ({ children }: any) => {
  return <ThemeProvider theme={dark}>{children}</ThemeProvider>;
};

const customRender = (ui: ReactElement, options?: any) =>
  render(ui, { wrapper: AllProviders, ...options });

export * from "@testing-library/react";

export { customRender as render, screen, dark as defaultTheme };
