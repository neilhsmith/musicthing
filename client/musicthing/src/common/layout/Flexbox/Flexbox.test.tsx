import { render } from "test-utils";

import Flexbox from "./Flexbox";

describe("Component: Flexbox", () => {
  it("renders correctly", () => {
    const { container } = render(
      <Flexbox
        alignContent="flex-start"
        alignItems="center"
        alignSelf="center"
        display="flex"
        element="article"
        flex="1"
        flexBasis="1"
        flexDirection="column"
        flexFlow="row"
        flexGrow="1"
        flexShrink="1"
        flexWrap="nowrap"
        justifyContent="space-around"
        margin="1rem"
        marginTop="1rem"
        marginRight="1rem"
        marginBottom="1rem"
        marginLeft="1rem"
        padding="1rem"
        paddingTop="1rem"
        paddingRight="1rem"
        paddingBottom="1rem"
        paddingLeft="1rem"
        height="1rem"
        maxHeight="1rem"
        minHeight="1rem"
        width="1rem"
        maxWidth="1rem"
        minWidth="1rem"
      >
        <span data-testid="42">a child</span>
      </Flexbox>
    );

    expect(container.firstChild).toMatchSnapshot();

    expect(container.firstChild).toHaveStyleRule("align-content", "flex-start");
    expect(container.firstChild).toHaveStyleRule("align-items", "center");
    expect(container.firstChild).toHaveStyleRule("align-self", "center");
    expect(container.firstChild).toHaveStyleRule("display", "flex");
    expect(container.firstChild).toHaveStyleRule("flex", "1");
    expect(container.firstChild).toHaveStyleRule("flex-basis", "1");
    expect(container.firstChild).toHaveStyleRule("flex-direction", "column");
    expect(container.firstChild).toHaveStyleRule("flex-flow", "row");
    expect(container.firstChild).toHaveStyleRule("flex-grow", "1");
    expect(container.firstChild).toHaveStyleRule("flex-shrink", "1");
    expect(container.firstChild).toHaveStyleRule("flex-wrap", "nowrap");
    expect(container.firstChild).toHaveStyleRule(
      "justify-content",
      "space-around"
    );
    expect(container.firstChild).toHaveStyleRule("margin", "1rem");
    expect(container.firstChild).toHaveStyleRule("margin-top", "1rem");
    expect(container.firstChild).toHaveStyleRule("margin-right", "1rem");
    expect(container.firstChild).toHaveStyleRule("margin-bottom", "1rem");
    expect(container.firstChild).toHaveStyleRule("margin-left", "1rem");
    expect(container.firstChild).toHaveStyleRule("padding", "1rem");
    expect(container.firstChild).toHaveStyleRule("padding-top", "1rem");
    expect(container.firstChild).toHaveStyleRule("padding-right", "1rem");
    expect(container.firstChild).toHaveStyleRule("padding-bottom", "1rem");
    expect(container.firstChild).toHaveStyleRule("padding-left", "1rem");
    expect(container.firstChild).toHaveStyleRule("height", "1rem");
    expect(container.firstChild).toHaveStyleRule("max-height", "1rem");
    expect(container.firstChild).toHaveStyleRule("min-height", "1rem");
    expect(container.firstChild).toHaveStyleRule("width", "1rem");
    expect(container.firstChild).toHaveStyleRule("max-width", "1rem");
    expect(container.firstChild).toHaveStyleRule("min-width", "1rem");
  });
});
