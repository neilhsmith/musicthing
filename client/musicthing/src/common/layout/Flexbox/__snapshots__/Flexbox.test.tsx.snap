// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Flexbox renders correctly 1`] = `
.c0 {
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-basis: 1;
  -ms-flex-preferred-size: 1;
  flex-basis: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-flow: row;
  -ms-flex-flow: row;
  flex-flow: row;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  margin: 1rem;
  margin-top: 1rem;
  margin-right: 1rem;
  margin-bottom: 1rem;
  margin-left: 1rem;
  padding: 1rem;
  padding-top: 1rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  padding-left: 1rem;
  height: 1rem;
  max-height: 1rem;
  min-height: 1rem;
  width: 1rem;
  max-width: 1rem;
  min-width: 1rem;
}

<article
  class="c0"
>
  <span
    data-testid="42"
  >
    a child
  </span>
</article>
`;
